lib/class-wp-webfonts.php

```diff
@@ -0,0 +1,323 @@
+
+		// Enqueue webfonts in the block editor.
+		add_action( 'admin_init', array( $this, 'generate_and_enqueue_editor_styles' ) );
+	}
+
+	/**
+	 * Get the list of fonts.
+	 *
+	 * @return array
+	 */
+	public function get_fonts() {
+		return self::$webfonts;
+	}
+
+	/**
+	 * Get the list of providers.
+	 *
+	 * @return array
+	 */
+	public function get_providers() {
+		return self::$providers;
+	}
+
+	/**
+	 * Register a webfont.
+	 *
+	 * @param array $font The font arguments.
+	 */
+	public function register_font( $font ) {
+		$font = $this->validate_font( $font );
+		if ( $font ) {
+			$id                    = $this->get_font_id( $font );
+			self::$webfonts[ $id ] = $font;
+		}
+	}
+
+	/**
+	 * Get the font ID.
+	 *
+	 * @param array $font The font arguments.
+	 * @return string
+	 */
+	public function get_font_id( $font ) {
+		return sanitize_title( "{$font['font-family']}-{$font['font-weight']}-{$font['font-style']}-{$font['provider']}" );
+	}
+
+	/**
+	 * Validate a font.
+	 *
+	 * @param array $font The font arguments.
+	 *
+	 * @return array|false The validated font arguments, or false if the font is invalid.
+	 */
+	public function validate_font( $font ) {
+		$font = wp_parse_args(
+			$font,
+			array(
+				'provider'     => 'local',
+				'font-family'  => '',
+				'font-style'   => 'normal',
+				'font-weight'  => '400',
+				'font-display' => 'fallback',
+			)
+		);
+
+		// Check the font-family.
+		if ( empty( $font['font-family'] ) || ! is_string( $font['font-family'] ) ) {
+			trigger_error( __( 'Webfont font family must be a non-empty string.', 'gutenberg' ) );
+			return false;
+		}
+
+		// Local fonts need a "src".
+		if ( 'local' === $font['provider'] ) {
+			// Make sure that local fonts have 'src' defined.
+			if ( empty( $font['src'] ) || ( ! is_string( $font['src'] ) && ! is_array( $font['src'] ) ) ) {
+				trigger_error( __( 'Webfont src must be a non-empty string or an array of strings.', 'gutenberg' ) );
+				return false;
+			}
+		}
+
+		// Validate the 'src' property.
+		if ( ! empty( $font['src'] ) ) {
+			foreach ( (array) $font['src'] as $src ) {
+				if ( empty( $src ) || ! is_string( $src ) ) {
+					trigger_error( __( 'Each webfont src must be a non-empty string.', 'gutenberg' ) );
+					return false;
+				}
+
+				if (
+					// Validate data URLs.
+					! preg_match( '/^data:.+;base64/', $src ) &&
+					// Validate URLs.
+					! filter_var( $src, FILTER_VALIDATE_URL ) &&
+					// Check if it's a URL starting with "//" (omitted protocol).
+					0 !== strpos( $src, '//' )
+				) {
+					trigger_error( __( 'Webfont src must be a valid URL or a data URI.', 'gutenberg' ) );
+					return false;
+				}
+			}
+		}
+
+		// Check the font-style.
+		$valid_font_styles = array( 'normal', 'italic', 'oblique', 'inherit', 'initial', 'revert', 'unset' );
+		if ( ! in_array( $font['font-style'], $valid_font_styles, true ) && ! preg_match( '/^oblique\s+(\d+)%/', $font['font-style'] ) ) {
+			return false;
+		}
```

Comment:
Not sure about this check too. What happens if `$font['font-style']` is invalid? It fails silently and the user is left wandering what's going on :) 
Perhaps better to allow the invalid string so the user can see an error in the browser console.

Reply:
Done 👍